_drafts/faster-builds.md

```diff
@@ -0,0 +1,649 @@
+- Ensure no timestamps are injected into the build which would force rebuilds.
+- Try not to enable link-time optimizations (LTO) as the linking step sometimes
+  becomes as slow as the entire build itself! If you have enabled it due to code
+  size constraints, I suggest checking out Interrupt's
+  [code size posts](https://interrupt.memfault.com/blog/tag/fw-code-size) for
+  easy wins.
+
+## Example Build Environment
+
+To put these suggestions to the test, I took the example
+LwIP_HTTP_Server_Netconn_RTOS from STM32CubeF4 project[^4], found a GCC
+port[^10], and used that to test the various operating systems, compilers, and
+build systems. To easily get the same environment locally, one can clone the
+project from the Interrupt repository and perform the appropriate steps for each
+compiler.
+
+```
+$ git clone git@github.com:memfault/interrupt.git
+$ cd interrupt/example/faster-compilation
+```
+
+> Make will automatically download the STM32CubeF4 and extract it to the
+> appropriate place. When not using Make, download the distribution from
+> [Github](https://github.com/STMicroelectronics/STM32CubeF4) and extract it to
+> `stm32cube/`
+
+- GCC - Using Make, run `make`. Works on macOS, Linux, and Windows
+- Keil - Open
+  `stm32cube/Projects/STM32F429ZI-Nucleo/Applications/LwIP/LwIP_HTTP_Server_Netconn_RTOS/MDK-ARM/Project.uvprojx`
+- IAR - Open
+  `stm32cube/Projects/STM32F429ZI-Nucleo/Applications/LwIP/LwIP_HTTP_Server_Netconn_RTOS/EWARM/Project.eww`
+
+Below are the versions of the programs tested:
+
+- Keil: MDK-Arm 5.29 on Windows, where `armcc` is ARM Compiler 5.06, and
+  `armclang` is ARM Compiler 6.13
+- IAR: IAR Embedded Workbench for ARM 8.42.1 on Windows
+- GCC: GNU ARM Embedded Toolchain 8-2019-q3[^9] for macOS, Linux, and Windows
+
+The operating system environments were as follows:
+
+- Windows: a VirtualBox image of Windows 10 was used, with 8GB RAM and 4 CPU
+- Linux: Ubuntu 18.04 on Docker, with 8GB RAM and 4 CPU
+- macOS: 10.14 Mojave native installation, 16GB RAM and 4 CPU.
+
+## Steps Towards Faster Builds
+
+### Parallel Compilation
+
+The first and easiest thing to do to speed up builds is ensure you are building
+with multiple threads. Make, IAR, and Keil all support parallel compilation, so
+ensure you and the entire team have enabled this.
+
+Below are some times recorded compiling the example project with GCC and
+different thread counts.
+
+![comparison of compiler build times with GCC threads](/img/faster-compilation/gcc-threads-build-time-comparison.svg)
+
+With Make, it is as simple as invoking Make with the argument `-jN`, where `N`
+is the number of desired threads to use. If you or your team members
+occasionally forget to pass in this parameter, I suggest wrapping the command
+[using
+Invoke]({% post_url 2019-08-27-building-a-cli-for-firmware-projects %}#adding-parallel-builds)
+
+If you are using Make in a recursive manner, **make sure** you are using
```

Comment:
```suggestion
If you are calling Make from a Makefile, verify that you are using
```

- Wasn't sure what you meant with "recursive manner" with it until I saw the example.
- The change to "verify" because there's too much "make" it that sentence already.
- I would drop the emphasis on just verbs without objects (there are other cases in this post) as it just adds visual noise. If you really need it, consider emphasising the whole action (similarly to how you should do it for links):

> If you are calling Make from a Makefile, **verify that you are using `${MAKE)` instead of `make`** when using it in a rule

Reply:
You are right. I need to adjust how I transcribe about these things from my mind and instead use better written language.